File: actojat-transpiler/src/main/java/de/netherspace/apps/actojat/util/SourceErrorListener.java
package de.netherspace.apps.actojat.util;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.antlr.v4.runtime.ANTLRErrorListener;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;

import java.util.BitSet;

/**
 * A Custom ErrorListener that sets a certain error flag whenever an error occurs.
 */
@Slf4j
public class SourceErrorListener extends BaseErrorListener implements ANTLRErrorListener {

  @Getter
  private boolean errorFlag;

  /**
   * The default constructor.
   */
  public SourceErrorListener() {
    super();
    this.errorFlag = false;
  }

  @Override
  public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol,
                          int line, int charPositionInLine,
                          String msg, RecognitionException exception) {
    this.errorFlag = true;
    log.error("Syntax Error found!" + msg);
    super.syntaxError(recognizer, offendingSymbol, line, charPositionInLine, msg, exception);
  }

  @Override
  public void reportAmbiguity(Parser recognizer, DFA dfa, int startIndex,
                              int stopIndex, boolean exact,
                              BitSet ambigAlts, ATNConfigSet configs) {
    this.errorFlag = true;
    log.error("Ambiguity found!");
    super.reportAmbiguity(recognizer, dfa, startIndex, stopIndex, exact, ambigAlts, configs);
  }

}

File: actojat-transpiler/src/main/kotlin/de/netherspace/apps/actojat/util/SourceErrorListener.kt
package de.netherspace.apps.actojat.util

import org.antlr.v4.runtime.*
import org.antlr.v4.runtime.atn.ATNConfigSet
import org.antlr.v4.runtime.dfa.DFA
import org.slf4j.LoggerFactory
import java.util.*

/**
 * A Custom ErrorListener that sets an error flag whenever an error occurs.
 */
class SourceErrorListener : BaseErrorListener(), ANTLRErrorListener {

    private val log = LoggerFactory.getLogger(SourceErrorListener::class.java)

    private var errorFlag = false

    override fun syntaxError(recognizer: Recognizer<*, *>?, offendingSymbol: Any?, line: Int, charPositionInLine: Int, msg: String?, e: RecognitionException?) {
        errorFlag = true
        log.error("Syntax Error found:  $msg")
        super.syntaxError(recognizer, offendingSymbol, line, charPositionInLine, msg, e)
    }

    override fun reportAmbiguity(recognizer: Parser?, dfa: DFA?, startIndex: Int, stopIndex: Int, exact: Boolean, ambigAlts: BitSet?, configs: ATNConfigSet?) {
        errorFlag = true
        log.error("Ambiguity found! startIndex = $startIndex, stopIndex = $stopIndex")
        super.reportAmbiguity(recognizer, dfa, startIndex, stopIndex, exact, ambigAlts, configs)
    }

    fun isErrorFlag(): Boolean {
        return errorFlag
    }

}
